### PR TITLE
Tickets/DM-9528: Cleanup pybind11 code in meas_deblender

### PR DESCRIPTION
**Pull request by fred3m**
Apply the pybind11 coding style guidelines from RFC-281 and the updated developer guide

### CODE REVIEW

File: python/lsst/meas/deblender/SConscript [hunk 1]
```diff
@@ -1,3 +1,3 @@
 # -*- python -*-
 from lsst.sconsUtils import scripts
-scripts.BasicSConscript.pybind11(['baseline'])
+scripts.BasicSConscript.pybind11(['baselineUtils'])
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/SConscript]: Please use `addUnderscore=False` in this call; that will let you call the module just `baselineUtils` everywhere else (and leave us only one place to change after not-adding-underscores becomes the default).

File: python/lsst/meas/deblender/baseline.py [hunk 2]
```diff
@@ -204,7 +203,7 @@ def __init__(self, filterName, footprint, maskedImage, psf, psffwhm, avgNoise,
         # avgNoise is an estiamte of the average noise level for the image in this filter
         if avgNoise is None:
             stats = afwMath.makeStatistics(self.varimg, self.mask, afwMath.MEDIAN)
-            avgNoise = math.sqrt(stats.getValue(afwMath.MEDIAN))
+            avgNoise = np.sqrt(stats.getValue(afwMath.MEDIAN))
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baseline.py]: Should go in a separate commit at least.

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: Please use double quotes instead of angle brackets for these includes.

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+#include "lsst/meas/deblender/BaselineUtils.h"
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace lsst {
+namespace meas {
+namespace deblender {
+
+namespace {
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: Add newline (between namespace open and function declaration).

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                   &BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::getSignificantEdgePixels, "img"_a,
+                   "sfoot"_a, "thresh"_a);
+    // There appears to be an issue binding to a static const member of a templated type, so for now
+    // we just use the values constants
+    cls.attr("ASSIGN_STRAYFLUX") = py::cast(0x1);
+    cls.attr("STRAYFLUX_TO_POINT_SOURCES_WHEN_NECESSARY") = py::cast(0x2);
+    cls.attr("STRAYFLUX_TO_POINT_SOURCES_ALWAYS") = py::cast(0x4);
+    cls.attr("STRAYFLUX_R_TO_FOOTPRINT") = py::cast(0x8);
+    cls.attr("STRAYFLUX_NEAREST_FOOTPRINT") = py::cast(0x10);
+    cls.attr("STRAYFLUX_TRIM") = py::cast(0x20);
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: We should use the enums rather than repeat their values as integers, even if it's more verbose - don't want any chance of this getting out of sync with the C++.

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+namespace deblender {
+
+namespace {
+template <typename ImagePixelT, typename MaskPixelT = lsst::afw::image::MaskPixel,
+          typename VariancePixelT = lsst::afw::image::VariancePixel>
+void declareBaselineUtils(py::module& mod, const std::string& suffix) {
+    using MaskedImageT = lsst::afw::image::MaskedImage<ImagePixelT, MaskPixelT, VariancePixelT>;
+    using ImagePtrT = std::shared_ptr<lsst::afw::image::Image<ImagePixelT>>;
+    using FootprintPtrT = std::shared_ptr<lsst::afw::detection::Footprint>;
+
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: I'd recommend doing
```
using Class = BaseLineUtils<ImagePixelT, MaskPixelT, VariancePixelT>;
```
here as well; it'll save quite a few characters below.

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    py::class_<BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>> cls(
+            mod, ("BaselineUtils" + suffix).c_str());
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: It probably doesn't matter in this case, but we should be using `shared_ptr` holder by default in case downstream code decides it wants to start using `shared_ptr`s to this class.  This might make it a bit more readable if you also defined a
```
using PyClass = py::class_<Class, std::shared_ptr<Class>>;
```
typedef.

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                   "img"_a, "outimg"_a, "halfsize"_a);
+    cls.def_static("makeMonotonic", &BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::makeMonotonic,
+                   "img"_a, "pk"_a);
+    // apportionFlux expects an empty vector containing HeavyFootprint pointers that is modified
+    // in the function. But when a list is passed to pybind11 in place of the vector,
+    // the changes are not passed back to python. So instead we create the vector in this lambda and
+    // include it in the return value.
+    cls.def_static("apportionFlux",
+                   [](MaskedImageT const& img, lsst::afw::detection::Footprint const& foot,
+                      std::vector<typename std::shared_ptr<lsst::afw::image::Image<ImagePixelT>>> templates,
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: No need for `typename` here (and some compilers might consider it illegal); you only need `typename` if you're doing `::<something>` on a templated type.

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                      std::vector<std::shared_ptr<lsst::afw::detection::Footprint>> templ_footprints,
+                      ImagePtrT templ_sum, std::vector<bool> const& ispsf, std::vector<int> const& pkx,
+                      std::vector<int> const& pky, int strayFluxOptions, double clipStrayFluxFraction) {
+                       using HeavyFootprintPtrList =
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: Weird indentation on this line.

Overall, I think this would be a lot more readable if you formatted as:
```
cls.def_static(
    "apportionFlux",
    [](
        MaskedImageT const& img,
        afw::detection::Footprint const& foot,
        ...
    ) {
        using HeavyFootprintList = ...
    }
);
```
That is, put each arg on a new line and put each brace- or open-paren on a new line.

You can of course use clang-format instead (which will probably not make it look as nice, but it should at least fix up the indentation).

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+                               std::vector<std::shared_ptr<typename lsst::afw::detection::HeavyFootprint<
+                                       ImagePixelT, MaskPixelT, VariancePixelT>>>;
+
+                       std::vector<typename std::shared_ptr<
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: More unneeded `typename`s here (and a few lines above).

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    py::module::import("lsst.afw.detection");
+    py::module mod("_baselineUtils", "Python wrapper for meas_deblender _baselineUtils library");
+
+    declareBaselineUtils<float>(mod, "F");
+
+    return mod.ptr();
+}
+}
+}
+}  // lsst::meas::deblender
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: Probably better to put all of these namespace closes on a single line.  Also add a newline after the `PYBIND11_PLUGIN { ... }` block close and before the first namespace close.

File: python/lsst/meas/deblender/baselineUtils.cc [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    cls.attr("STRAYFLUX_R_TO_FOOTPRINT") = py::cast(0x8);
+    cls.attr("STRAYFLUX_NEAREST_FOOTPRINT") = py::cast(0x10);
+    cls.attr("STRAYFLUX_TRIM") = py::cast(0x20);
+};
+}  // namespace
+
+PYBIND11_PLUGIN(_baselineUtils) {
+    py::module::import("lsst.afw.image");
+    py::module::import("lsst.afw.detection");
+    py::module mod("_baselineUtils", "Python wrapper for meas_deblender _baselineUtils library");
```
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: You can (and probably should) drop the docstring here.  It doesn't actually add any information.

File: python/lsst/meas/deblender/baselineUtils.cc (hunk 9 above)
Comment (fred3m) [inline python/lsst/meas/deblender/baselineUtils.cc]: I did run clang format, and I'm pretty sure that is what used this indentation (for whatever reason). I'll see if re-runnning it will fix things.

File: python/lsst/meas/deblender/baselineUtils.cc (hunk 11 above)
Comment (fred3m) [inline python/lsst/meas/deblender/baselineUtils.cc]: They were originally, clang-format did this.

File: python/lsst/meas/deblender/baselineUtils.cc (hunk 5 above)
Comment (fred3m) [inline python/lsst/meas/deblender/baselineUtils.cc]: So you want me to change the C++ code, this is not an enum on the C++ side right now?

File: python/lsst/meas/deblender/baselineUtils.cc (hunk 9 above)
Comment (fred3m) [inline python/lsst/meas/deblender/baselineUtils.cc]: Hmmm, maybe it's because I used ``Class`` instead of ``BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>`` but clang-format did a much better job this time.

File: python/lsst/meas/deblender/baselineUtils.cc (hunk 5 above)
Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: Oh, I didn't realize it wasn't an enum.  That's actually not what I was getting at, though; I was asking for wrappers like:
```
cls.attr("STRAYFLUX_TRIM") = py::cast(BaselineUtils::STRAYFLUX_TRIM);
```
I just to avoid actually defining the numbers themselves in multiple places.

Comment (fred3m) [inline python/lsst/meas/deblender/baselineUtils.cc]: That's how I tried to wrap them but there is a pybind11 bug that prevents wrapping a constant member of a static type. That's why I included the comment 

```
// There appears to be an issue binding to a static const member of a templated type, so for now
// we just use the constants values
```

before the code block. @pschella and I discussed this when I was wrapping the module and while neither of us liked this solution, it seemed like the best way to handle this for now.

Comment (TallJimbo) [inline python/lsst/meas/deblender/baselineUtils.cc]: It's a bit tricky, but adding the following to `src/BaselineUnits.cc` before the explicit instantiation will work:
```
template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT>
const int deblend::BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::ASSIGN_STRAYFLUX;

template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT>
const int
deblend::BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::STRAYFLUX_TO_POINT_SOURCES_WHEN_NECESSARY;

template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT>
const int deblend::BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::STRAYFLUX_TO_POINT_SOURCES_ALWAYS;

template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT>
const int deblend::BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::STRAYFLUX_R_TO_FOOTPRINT;

template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT>
const int deblend::BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::STRAYFLUX_NEAREST_FOOTPRINT;

template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT>
const int deblend::BaselineUtils<ImagePixelT, MaskPixelT, VariancePixelT>::STRAYFLUX_TRIM;
```
Full commit is on branch u/jbosch/DM-9528.

Comment (fred3m) [inline python/lsst/meas/deblender/baselineUtils.cc]: Aha, that did work. 😄

File: python/lsst/meas/deblender/baseline.py (hunk 2 above)
Comment (fred3m) [inline python/lsst/meas/deblender/baseline.py]: Agreed, but at this point it seems like a lot of trouble for such a trivial change. Would you be ok with not worrying about splitting the commits and leaving it as is? If so, this branch is ready to merge.

Comment (TallJimbo) [inline python/lsst/meas/deblender/baseline.py]: Okay with me.
